src/vos/vos_obj.c

```diff
@@ -1928,26 +1937,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
 	krec = rbund.rb_krec;
 	umm = vos_obj2umm(oiter->it_obj);
 
-	if (!vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
-				 &oiter->it_punched, &oiter->it_ilog_info))
+	punched = vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
+				      &oiter->it_punched, &oiter->it_ilog_info);
+
+	bmap = krec->kr_bmap;
+
+	if (bmap & KREC_BF_AGG_OPT) {
+		if ((bmap & KREC_BF_AGG_NEEDED) == 0) {
+			if (!punched && oiter->it_punched.pr_epc == 0)
+				return 2;
+		} else {
+			bmap |= KREC_BF_AGG_FLAG;
```

Comment:
Introduce an extra transaction to for each key to set KREC_BF_AGG_FLAG flag looks overkilled to me.

Reply:
how would you handle it otherwise?

Reply:
Like tolerating a full scan after server restart and update flags without persistence guarantee. I'm afraid that flushing and invalidating CPU cache when iterating each entry could impact performance, but let's evaluate the performance first then make decision.

Reply:
Doesn't cache invalidation only happen with Apache Pass?  I thought they were fixing that in later renditions.